File: core/lib/asset_map/asset_map.js
/**
* +--------------------------------------------------------------------+
* | Squiz.net Open Source Licence                                      |
* +--------------------------------------------------------------------+
* | Copyright (c), 2003 Squiz Pty Ltd (ABN 77 084 670 600).            |
* +--------------------------------------------------------------------+
* | This source file may be used subject to, and only in accordance    |
* | with, the Squiz Open Source Licence Agreement found at             |
* | http://www.squiz.net/licence.                                      |
* | Make sure you have read and accept the terms of that licence,      |
* | including its limitations of liability and disclaimers, before     |
* | using this software in any way. Your use of this software is       |
* | deemed to constitute agreement to be bound by that licence. If you |
* | modify, adapt or enhance this software, you agree to assign your   |
* | intellectual property rights in the modification, adaptation and   |
* | enhancement to Squiz Pty Ltd for use and distribution under that   |
* | licence.                                                           |
* +--------------------------------------------------------------------+
*
* $Id: asset_map.js,v 1.12 2005/02/18 05:30:32 mmcintyre Exp $
*
*/


var SQ_REFRESH_ASSETIDS = "";


/**
* Initalises the asset map
*
* Checks the browser flavour and version and make some desicions based on that information
* about whether communication should be stable between the javascript and the asset map
*/
function init_asset_map()
{
	if (matches = navigator.userAgent.match(/MSIE ([0-9.]+)/)) {

		if (matches[1] < '6.0') {
			alert('You need to use Internet Explorer 6.0 or above for the communication between the Asset Map and the Javascript');
		}

	} else {

		var firebird_re = /(Firebird)\/([0-9.]+)/;
		var safari_re   = /(Safari)\/([0-9.]+)/;
		var moz_re      = /^Mozilla\/5\.0.*rv:([^)]+)\)/;

		if ((matches = navigator.userAgent.match(firebird_re)) || (matches = navigator.userAgent.match(safari_re)) || (matches = navigator.userAgent.match(moz_re))) {
			if (matches[1] == 'Firebird' && matches[2] < '0.6.1') {
				alert('You need to use Firebird 0.6.1 or above for the communication between the Asset Map and the Javascript');
			} else if (matches[1] == 'Safari' && parseFloat(matches[2]) < 125.1) {
				// Safari 1.2.1 reports as build number '125.1'
				alert('You need to use Safari 1.2.1 (build 125.1) or above for the communication between the Asset Map and the Javascript');
			} else if (matches[1] == 'Mozilla' && matches[2] < '1.4') {
				alert('You need to use Mozilla 1.4 or above for the communication between the Asset Map and the Javascript');
			}
		// we don't know about this browser, ah well may as well give it a go ...
		} else {
			alert('You are using an untested browser there is no guarantee that the communication between the Asset Map and the Javascript will be successful');
		}

	}//end if

}//end init_asset_map()


/**
* Opens a new window with a HIPO job in it
*
*/
function open_hipo(url)
{
	window.open(url, 'hipo_job', 'width=650,height=400,scrollbars=1,toolbar=0,menubar=0,location=0,resizable=1');

}//end open_hipo()


/**
* Returns the java applet object
*
* @return &object the java applet object
*/
function get_java_applet_object()
{
	return document.sq_asset_map;

}//end get_java_applet_object()


/**
* Add the passed list of assetids to the array of IDs to refresh
*
*/
function reload_assets(assetids)
{
	if (SQ_REFRESH_ASSETIDS != "") {
		SQ_REFRESH_ASSETIDS += ",";
	}
	SQ_REFRESH_ASSETIDS += assetids;

}//end reload_assets()


/*
* Add the passed assetid to the array of IDs to refresh
*
*/
function reload_asset(assetid)
{
	if (SQ_REFRESH_ASSETIDS != "") {
		SQ_REFRESH_ASSETIDS += ",";
	}
	SQ_REFRESH_ASSETIDS += assetid;

}//end reload_asset()


/**
* Makes an extenal call to java
*
* @param object		$asset_mapObj	the java applet
* @param string		$type			the type of request
* @param String		$command		the command
* @param Array		$params			the params to pass to java
*/
function jsToJavaCall(asset_mapObj, type, command, params)
{
	
	var params_str = '';
	for (var i = 0; i < params.length; i++) {
		params_str += params[i];
		if (params.length - 1 != i) {
			params_str += ',';
		}
	}
	asset_mapObj.jsToJavaCall(type, command, params_str);

}//end jsToJavaCall();




  //////////////////////////////////
 //   Asset Finder Functions     //
//////////////////////////////////

var ASSET_FINDER_FIELD_NAME = null;
var ASSET_FINDER_FIELD_SAFE_NAME = null;
var ASSET_FINDER_DONE_FUNCTION = null;
var ASSET_FINDER_OBJ = null;

/**
* set the finder object that initiated the asset finder
*
* @param finder the finder that initiated the asset finder
*/
function set_finder(finder)
{
	ASSET_FINDER_OBJ = finder;

}//end set_finder()


/**
* Activated by the pressing of the "Change" button to start the asset finder mode in the flash menu
*
* @param string	$name			the name of the hidden field
* @param string	$safe_name		the name prefix for all the other form elements associated with the
* @param string	$type_codes		the typecodes seperated by a pipe
*
* @access public
*/
function asset_finder_change_btn_press(name, safe_name, type_codes, done_fn)
{
	ASSET_FINDER_DONE_FUNCTION = done_fn;

	if (ASSET_FINDER_FIELD_NAME != null && ASSET_FINDER_FIELD_NAME != name) {
		alert('The asset finder is currently in use');
		return;
	}

	// no name ? we must be starting the asset finder
	if (ASSET_FINDER_FIELD_NAME == null) {
		ASSET_FINDER_FIELD_NAME = name;
		ASSET_FINDER_FIELD_SAFE_NAME = safe_name;

		asset_finder_start('asset_finder_done', type_codes);

		ASSET_FINDER_OBJ.set_button_value(ASSET_FINDER_FIELD_SAFE_NAME + '_change_btn', 'Cancel');

	// else we must be cancelling the asset finder
	} else {
		asset_finder_cancel();
		ASSET_FINDER_OBJ.set_button_value(ASSET_FINDER_FIELD_SAFE_NAME + '_change_btn', 'Change');
		ASSET_FINDER_FIELD_NAME = null;
		ASSET_FINDER_FIELD_SAFE_NAME = null;
	}

}//end asset_finder_change_btn_press()


/**
* Call-back fns that stops the asset finder
*
* @param Array params the params array
* @param string label the label to give this asset
* @param string url the url of this asset
*
* @access public
*/
function asset_finder_done(params, label, url)
{
	var win = ASSET_FINDER_OBJ.window;
	win.focus();

	if (ASSET_FINDER_FIELD_NAME == null) return;

	var assetid = params;
	// we are in java, params is actually the assetid
	if (assetid != -1) {
		ASSET_FINDER_OBJ.set_hidden_field(ASSET_FINDER_FIELD_NAME + '[assetid]',assetid);
		ASSET_FINDER_OBJ.set_hidden_field(ASSET_FINDER_FIELD_NAME + '[url]', url);
		ASSET_FINDER_OBJ.set_text_field(ASSET_FINDER_FIELD_SAFE_NAME + '_label', (assetid == 0) ? '' : label + ' (Id : #' + assetid + ')');

	}

	ASSET_FINDER_OBJ.set_button_value(ASSET_FINDER_FIELD_SAFE_NAME + '_change_btn', 'Change');
	ASSET_FINDER_FIELD_NAME = null;
	ASSET_FINDER_FIELD_SAFE_NAME = null;
	if (ASSET_FINDER_DONE_FUNCTION !== null) ASSET_FINDER_DONE_FUNCTION();

}//end asset_finder_done()


/**
* Starts the asset finder and lets the asset map know that we are now in asset finder mode
*
* @param String		fn				the function to call when done
* @param String		type_codes		the type codes to restrict the asset finder to
*
* @access public
*/
var ASSET_FINDER_CALL_BACK = null;
function asset_finder_start(fn, type_codes)
{
	var params = type_codes.split('|');
	var asset_mapObj = get_java_applet_object();
	jsToJavaCall(asset_mapObj, 'asset_finder', 'assetFinderStarted', params);

}//end asset_finder_start()


/**
* Alerts the asset map that asset finder mode has been canceled
*
* @access public
*/
function asset_finder_cancel() {

	var asset_mapObj = get_java_applet_object();
	params = new Array();
	jsToJavaCall(asset_mapObj, 'asset_finder', 'assetFinderStopped', params);

}//end asset_finder_cancel()


/**
* Activated by the pressing of the "Clear" button
*
* @param string	$name			the name of the hidden field
* @param string	$safe_name		the name prefix for all the other form elements associated with the
*
* @access public
*/
function asset_finder_clear_btn_press(name, safe_name)
{
	ASSET_FINDER_OBJ.set_hidden_field(name + '[assetid]', 0);
	ASSET_FINDER_OBJ.set_hidden_field(name + '[url]', '');
	ASSET_FINDER_OBJ.set_text_field(safe_name + '_label', '');

}//end asset_finder_clear_btn_press()


/**
* Activated by the pressing of the "Reset" button
*
* @param string	$name			the name of the hidden field
* @param string	$safe_name		the name prefix for all the other form elements associated with the
*
* @access public
*/
function asset_finder_reset_btn_press(name, safe_name, assetid, label)
{
	ASSET_FINDER_OBJ.set_hidden_field(name + '[assetid]', assetid);
	ASSET_FINDER_OBJ.set_text_field(safe_name + '_label', label);

}//end asset_finder_reset_btn_press()


/**
* Activated by on an unload event to cancel the asset finder if we are currently looking
*
* @access public
*/
function asset_finder_onunload()
{
	// got a name ? we must be finding assets, cancel it
	if (ASSET_FINDER_FIELD_NAME != null) {
		asset_finder_cancel();
	}

}//end asset_finder_onunload()


ASSET_FINDER_OTHER_ONUNLOAD = (window.onunload) ? window.onunload : new Function;
window.onunload = asset_finder_onunload;

File: core/lib/asset_map/java/src/net/squiz/matrix/assetmap/AssetMap.java
/**
* +--------------------------------------------------------------------+
* | Squiz.net Open Source Licence                                      |
* +--------------------------------------------------------------------+
* | Copyright (c), 2003 Squiz Pty Ltd (ABN 77 084 670 600).            |
* +--------------------------------------------------------------------+
* | This source file may be used subject to, and only in accordance    |
* | with, the Squiz Open Source Licence Agreement found at             |
* | http://www.squiz.net/licence.                                      |
* | Make sure you have read and accept the terms of that licence,      |
* | including its limitations of liability and disclaimers, before     |
* | using this software in any way. Your use of this software is       |
* | deemed to constitute agreement to be bound by that licence. If you |
* | modify, adapt or enhance this software, you agree to assign your   |
* | intellectual property rights in the modification, adaptation and   |
* | enhancement to Squiz Pty Ltd for use and distribution under that   |
* | licence.                                                           |
* +--------------------------------------------------------------------+
*
* $Id: AssetMap.java,v 1.12 2005/03/06 22:13:26 mmcintyre Exp $
*
*/

package net.squiz.matrix.assetmap;

import javax.swing.*;
import javax.swing.tree.*;
import net.squiz.matrix.ui.*;
import net.squiz.matrix.core.*;
import net.squiz.matrix.matrixtree.*;
import net.squiz.matrix.plaf.*;
import java.util.*;
import java.awt.event.*;
import java.awt.*;
import java.net.*;
import netscape.javascript.*;
import javax.swing.plaf.metal.*;
import javax.swing.border.*;
import net.squiz.matrix.debug.*;
import java.io.IOException;

/**
 * The main applet class
 * @author Marc McIntyre <mmcintyre@squiz.net>
 */
public class AssetMap extends JApplet implements InitialisationListener {

	private BasicView view1;
	private BasicView view2;
	private MatrixTabbedPane pane;
	protected static JSObject window;
	public static AssetMap applet;
	private javax.swing.Timer timer;
	public static final int POLLING_TIME = 2000;
	
	/**
	 * Constructs the Asset Map
	 */
	public AssetMap() {
		try {
			UIManager.setLookAndFeel(new MatrixLookAndFeel());
		} catch (UnsupportedLookAndFeelException ulnfe) {
			ulnfe.printStackTrace();
		}
		applet = this;
	}

	// MM: find a better solution for doing this
	// there is a way to get the root container (this)
	public static AssetMap getApplet() {
		return applet;
	}

	/**
	 * Opens the specfied url in the sq_main frame of the matrix system.
	 * @param url the url to open in the sq_main frame
	 */
	public static void getURL(String url) throws MalformedURLException {
		applet.getAppletContext().showDocument(new URL(url), "sq_main");
	}

	/**
	 * Operns a new browser window to the specified url
	 * @param url the url to open
	 * @param title the title to show in the browser window
	 * @see getURL(String)
	 */
	public static void openWindow(String url, String title) {
		if (window == null)
			window = JSObject.getWindow(applet);
		window.call("open_hipo", new Object[] { url } );
	}

	public void initialisationComplete(InitialisationEvent evt) {}

	
	/**
	 * Initialises the asset map.
	 * @see start()
	 * @see init()
	 */
	public void init() {
		window = JSObject.getWindow(this);
		loadParameters();
		getContentPane().add(createApplet());
	}

	/**
	 * Starts the applet.
	 * This is where the initial request is made to the matrix system for
	 * the current assets and asset types. The request happens in a swing safe
	 * worker thread and the GUI is updated upon completion of the request.
	 * @see JApplet.start()
	 * @see stop()
	 * @see init()
	 */
	public void start() {
		initAssetMap();
		startPolling();
	}
	
	/**
	 * Initialises the asset map by making a request to the matrix system for
	 * the current assets and asset types in the system.
	 */
	protected void initAssetMap() {
		// get a swing worker to call init in AssetManager
		// when it returns set the root node to all trees
		MatrixStatusBar.setStatus("Initialising...");
		SwingWorker worker = new SwingWorker() {
			public Object construct() {
				MatrixTreeNode root = null;
				try {
					root = AssetManager.init();
				} catch (IOException ioe) {
					return ioe;
				}
				return root;
			}
			
			public void finished() {
				Object get = get();
				// success
				if (get instanceof MatrixTreeNode) {
					MatrixTreeModelBus.setRoot((MatrixTreeNode) get());
					MatrixStatusBar.setStatusAndClear("Success!", 1000);
				// error
				} else if (get instanceof IOException) {
					IOException ioe = (IOException) get;
					GUIUtilities.error(
						AssetMap.this, ioe.getMessage(), "Initilisation Failed!");
					MatrixStatusBar.setStatusAndClear("Initilisation Failed!", 1000);
					Log.log("Could not initialise the asset map", AssetMap.class, ioe);
				}
			}
		};
		worker.start();
	}
	
	/**
	 * Starts the polling operation to refresh assets that Matrix deems stale.
	 * The polling operation is polled at POLLING_TIME intervals. When javascript
	 * array is not empty, the assetids are used in a refresh operation.
	 * @see POLLING_TIME
	 */
	protected void startPolling() {
		// Polling timer
		ActionListener taskPerformer = new ActionListener() {
			public void actionPerformed(ActionEvent evt) {
				if (window == null)
					JSObject.getWindow(AssetMap.this);

				String assetidsStr = (String) window.getMember("SQ_REFRESH_ASSETIDS");
				// if the string isn't empty, we have some work to do.
				// split the string for the asset ids and get the refresh worker
				// to start the refresh operation
				if (assetidsStr != null && !assetidsStr.equals("")) {
					String[] assetids = assetidsStr.split(",");
					AssetRefreshWorker worker = new AssetRefreshWorker(assetids, false) {
						// return a custom message for the wait message
						protected String getStatusBarWaitMessage() {
							return "Auto Refreshing...";
						}
					};
					worker.start();
					// clear the assets that we have refreshed
					window.eval("SQ_REFRESH_ASSETIDS = '';");
				}
			}
		};
		timer = new javax.swing.Timer(POLLING_TIME, taskPerformer);
		timer.start();
	}
	
	/**
	 * Load the parameters from the applet tag and add them to
	 * our property set. Parameters are stored in the Matrix property set
	 * and can be accessed with:
	 * <pre>
	 *  String prop = Matrix.getParameter("parameter." + propertyName);
	 * </pre>
	 * @see Matrix.getParameter(String)
	 */
	public void loadParameters() {
		// get the list of parameters availble to the asset map
		// and store them in the matrix property set.
		String paramStr = getParameter("parameter.params");
		String[] params = paramStr.split(",");

		for (int i = 0; i < params.length; i++)
			Matrix.setProperty(params[i], getParameter(params[i]));
	}

	/**
	 * Stops the asset map and polling.
	 * @see start()
	 * @see init()
	 */
	public void stop() {
		timer.stop();
		timer = null;
	}

	/**
	 * Creates the asset map's interface.
	 * @return the compoent to add to the content pane.
	 */
	protected JComponent createApplet() {
		getContentPane().setBackground(new Color(0x342939));

		pane = new MatrixTabbedPane(JTabbedPane.LEFT);
		view1 = new BasicView();
		view2 = new BasicView();

		pane.addView("Tree One", view1);
		pane.addView("Tree Two", view2);

		return pane;
	}

	/**
	 * Method stub for javascript calls to java.
	 * Passes on the work to the <code>JsEventManager</code> to invoke
	 * the method in the listeners.
	 *
	 * @param	type		The type of event
	 * @param	command		The event command. This is actually the
	 * name of the method that will eventually get  called in the listener.
	 * <br /><br />
	 * For Example.<br /> When javascript invokes the asset finder,
	 * a command of AssetFinderStarted is passed as this argument,
	 * which will call the method AssetFinderStarted in any listeners
	 * implementing the interface <code>AssetFinderListener</code>.
	 * @param params a serialised javascript array using the serialiseArray()
	 * function in <code>fudge/var_serialise/var_serialize.js</code>
	 */
	public void jsToJavaCall(String type, String command, String params) {
		if (type.equals("asset_finder")) {
			processAssetFinder(command, params);
		}
	}

	private void processAssetFinder(String command, String params) {
		if (command.equals("assetFinderStarted")) {
			String[] assetTypes = params.split(",");
			MatrixTreeBus.startAssetFinderMode(assetTypes);
		} else if (command.equals("assetFinderStopped")) {
			MatrixTreeBus.stopAssetFinderMode();
		}
	}
}
